# Preenchendo valores ausentes em DataFrames do Pandas

A função fillna() é usada para preencher valores ausentes

In [2]:
import pandas as pd
import numpy as np

In [3]:
# Importanto o dataset

dataSet = pd.read_csv("dataset.csv")


In [4]:
dataSet.head(5)

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         NaN  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         NaN  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [5]:
dataSet.isna().sum() # Verifica quantos valores não preenchidos existem por coluna

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      2
dtype: int64

In [6]:
moda = dataSet['Quantidade'].value_counts().index[0]

In [7]:
moda

3.0

In [8]:
# Após descobrir a 'moda' dos dados, vamos preencher o valor ausente com o valor

In [9]:
dataSet['Quantidade'].fillna(value = moda, inplace = True)

In [10]:
dataSet.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [11]:
valorTotal = dataSet['Valor_Venda'].value_counts().sum()

In [12]:
print(f"Valor total da tabela é R$ {valorTotal},00 ")

Valor total da tabela é R$ 9994,00 


In [13]:
recorrencia = dataSet['Segmento'].value_counts()

In [14]:
recorrencia

Consumer       5191
Corporate      3020
Home Office    1783
Name: Segmento, dtype: int64

## Query (Consulta)

In [15]:
dataSet.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [16]:
#mehtod query()

dataSet.Valor_Venda.describe()

count     9994.000000
mean       229.858001
std        623.245101
min          0.444000
25%         17.280000
50%         54.490000
75%        209.940000
max      22638.480000
Name: Valor_Venda, dtype: float64

O intervalo de vendas (Valor_Venda) é de 0.44 a 22638. Façamos consulta com retorno em todas as vendas dentro de um range de valores. 

In [17]:
intervaloVendas = dataSet.query('229 < Valor_Venda < 10000')

In [18]:
intervaloVendas

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
0     CA-2016-152156  2016-11-08   CG-12520     Consumer  United States   
1     CA-2016-152156  2016-11-08   CG-12520     Consumer  United States   
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
7     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
...              ...         ...        ...          ...            ...   
9973  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9976  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9979  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9991  CA-2017-121258  2017-02-26   DB-13060     Consumer  United States   
9993  CA-2017-119914  2017-05-04   CC-12220     Consumer  United States   

     Regiao       ID_Produto        Categoria  \
0     South  FUR-BO-10001798        Furniture   
1     South  FUR-CH-10000454        Furniture   
3     South  FUR-TA-10000577        Furniture   
7      West  TEC-PH-10002275       Technology   
10     West  FUR-TA-10001539        Furniture   
...     ...              ...              ...   
9973   West  TEC-PH-10004080       Technology   
9976   West  TEC-PH-10002496       Technology   
9979   West  OFF-BI-10002026  Office Supplies   
9991   West  TEC-PH-10003645       Technology   
9993   West  OFF-AP-10002684  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
0                     Bush Somerset Collection Bookcase     261.9600   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400   
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
7                        Mitel 5320 IP Phone VoIP phone     907.1520   
10             Chromcraft Rectangular Conference Tables    1706.1840   
...                                                 ...          ...   
9973                           Avaya 5410 Digital phone     271.9600   
9976                                       Cisco SPA301     249.5840   
9979                  Ibico Recycled Linen-Style Covers     437.4720   
9991                              Aastra 57i VoIP phone     258.5760   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...     243.1600   

      Quantidade  
0            3.0  
1            3.0  
3            5.0  
7            6.0  
10           9.0  
...          ...  
9973         5.0  
9976         2.0  
9979        14.0  
9991         2.0  
9993         2.0  

[2357 rows x 11 columns]

In [19]:
intervaloVendas.Valor_Venda.describe()

count    2357.000000
mean      766.679142
std       856.315136
min       229.544000
25%       323.100000
50%       490.320000
75%       859.200000
max      9892.740000
Name: Valor_Venda, dtype: float64

In [20]:
intervalo3 = intervaloVendas.query('Valor_Venda > 766')

In [21]:
intervalo3.describe()

Valor_Venda  Quantidade
count   687.000000  687.000000
mean   1615.597817    5.599709
std    1203.403998    2.498081
min     767.214000    1.000000
25%     914.200000    4.000000
50%    1212.960000    5.000000
75%    1763.575000    7.000000
max    9892.740000   14.000000

In [22]:
intervalo3.head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
3   US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
7   CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
10  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
11  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
24  CA-2015-106320  2015-09-25   EB-13870  Consumer  United States   West   

         ID_Produto   Categoria  \
3   FUR-TA-10000577   Furniture   
7   TEC-PH-10002275  Technology   
10  FUR-TA-10001539   Furniture   
11  TEC-PH-10002033  Technology   
24  FUR-TA-10000577   Furniture   

                                     Nome_Produto  Valor_Venda  Quantidade  
3   Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
7                  Mitel 5320 IP Phone VoIP phone     907.1520         6.0  
10       Chromcraft Rectangular Conference Tables    1706.1840         9.0  
11  Konftel 250 Conference phone - Charcoal black     911.4240         4.0  
24  Bretford CR4500 Series Slim Rectangular Table    1044.6300         3.0

## Verificando Ocorrências de valores em uma coluna

In [23]:
dataSet.shape

(9994, 11)

In [24]:
dataSet.ndim

2

In [25]:
filtroQtd2 = dataSet[dataSet['Quantidade'].isin([2])]

In [26]:
filtroQtd2.describe()

Valor_Venda  Quantidade
count  2401.000000      2401.0
mean    120.295510         2.0
std     252.773342         0.0
min       0.984000         2.0
25%      11.160000         2.0
50%      32.480000         2.0
75%     121.780000         2.0
max    4799.984000         2.0

In [27]:
# SHAPE

dataSet[dataSet['Quantidade'].isin([7])].shape

(606, 11)

In [28]:
dataSet[dataSet['Quantidade'].isin([7])][:10]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
5   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
21  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
22  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
27  US-2015-150630  2015-09-17   TB-21520     Consumer  United States   
35  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   
53  CA-2016-105816  2016-12-11   JM-15265    Corporate  United States   
83  CA-2015-149734  2015-09-03   JC-16105    Corporate  United States   
88  CA-2016-159695  2016-04-05   GM-14455  Home Office  United States   
96  CA-2017-161018  2017-11-09   PN-18775  Home Office  United States   
99  CA-2016-158568  2016-08-29   RB-19465  Home Office  United States   

     Regiao       ID_Produto        Categoria  \
5      West  FUR-FU-10001487        Furniture   
21  Central  OFF-AR-10000246  Office Supplies   
22  Central  OFF-AP-10001492  Office Supplies   
27     East  FUR-BO-10004834        Furniture   
35  Central  TEC-PH-10004977       Technology   
53     East  OFF-FA-10000304  Office Supplies   
83    South  OFF-EN-10000927  Office Supplies   
88  Central  OFF-ST-10003442  Office Supplies   
96     East  FUR-FU-10000629        Furniture   
99  Central  OFF-PA-10003256  Office Supplies   

                                         Nome_Produto  Valor_Venda  Quantidade  
5   Eldon Expressions Wood and Plastic Desk Access...       48.860         7.0  
21                                         Newell 318       19.460         7.0  
22        Acco Six-Outlet Power Strip, 4' Cord Length       60.340         7.0  
27  Riverside Palais Royal Lawyers Bookcase, Royal...     3083.430         7.0  
35                                        GE 30524EE4     1097.544         7.0  
53                                 Advantus Push Pins       15.260         7.0  
83      Jet-Pak Recycled Peel 'N' Seal Padded Mailers      200.984         7.0  
88                      Eldon Portable Mobile Manager      158.368         7.0  
96                    9-3/4 Diameter Round Wall Clock       96.530         7.0  
99         Avery Personal Creations Heavyweight Cards       64.624         7.0

## Operadores lógios para trabalho com PANDAS

In [29]:
dataSet[ (dataSet.Segmento == 'Home Office') & (dataSet.Regiao == 'South')]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
182   CA-2014-158274  2014-11-19   RM-19675  Home Office  United States   
183   CA-2014-158274  2014-11-19   RM-19675  Home Office  United States   
184   CA-2014-158274  2014-11-19   RM-19675  Home Office  United States   
231   US-2017-100930  2017-04-07   CS-12400  Home Office  United States   
232   US-2017-100930  2017-04-07   CS-12400  Home Office  United States   
...              ...         ...        ...          ...            ...   
9805  CA-2016-136322  2016-10-21   AP-10720  Home Office  United States   
9806  CA-2016-136322  2016-10-21   AP-10720  Home Office  United States   
9960  CA-2017-141446  2017-09-16   CL-12700  Home Office  United States   
9970  CA-2015-103772  2015-06-28   MP-17470  Home Office  United States   
9971  CA-2015-103772  2015-06-28   MP-17470  Home Office  United States   

     Regiao       ID_Produto        Categoria  \
182   South  TEC-PH-10003273       Technology   
183   South  TEC-PH-10004896       Technology   
184   South  TEC-AC-10002345       Technology   
231   South  FUR-TA-10001705        Furniture   
232   South  FUR-TA-10003473        Furniture   
...     ...              ...              ...   
9805  South  FUR-FU-10002878        Furniture   
9806  South  OFF-BI-10004817  Office Supplies   
9960  South  TEC-AC-10002305       Technology   
9970  South  OFF-BI-10002867  Office Supplies   
9971  South  OFF-AR-10000538  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
182                                        AT&T TR1909W     503.9600   
183                          Nokia Lumia 521 (T-Mobile)     149.9500   
184                   HP Standard 104 key PS/2 Keyboard      29.0000   
231    Bush Advantage Collection Round Conference Table     233.8600   
232          Bretford Rectangular Conference Table Tops     620.6145   
...                                                 ...          ...   
9805                Seth Thomas 14" Day/Date Wall Clock      45.5680   
9806                       GBC Personal VeloBind Strips      28.7520   
9960              KeyTronic E03601U1 - Keyboard - Beige      18.0000   
9970            GBC Recycled Regency Composition Covers     119.5600   
9971  Boston Model 1800 Electric Pencil Sharpener, Gray     140.7500   

      Quantidade  
182          4.0  
183          5.0  
184          2.0  
231          2.0  
232          3.0  
...          ...  
9805         2.0  
9806         8.0  
9960         1.0  
9970         2.0  
9971         5.0  

[272 rows x 11 columns]

In [30]:
filtro1 = dataSet[ (dataSet.Segmento == 'Home Office') & (dataSet.Regiao == 'South')]

In [31]:
filtro1['Valor_Venda'].sum()

74255.0015

In [32]:
filtro2 = dataSet [(dataSet.Segmento !='Home Office')]

In [33]:
filtro2.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [39]:
#ultimos registros

filtro3 = dataSet[(dataSet.Segmento == 'Corporate') | (dataSet.Regiao == 'South')].tail() 

In [40]:
filtro3

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
9971  CA-2015-103772  2015-06-28   MP-17470  Home Office  United States   
9980  US-2015-151435  2015-09-06   SW-20455     Consumer  United States   
9987  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9989  CA-2014-110422  2014-01-21   TB-21400     Consumer  United States   

     Regiao       ID_Produto        Categoria  \
9971  South  OFF-AR-10000538  Office Supplies   
9980  South  FUR-TA-10001029        Furniture   
9987  South  TEC-AC-10001539       Technology   
9988  South  TEC-PH-10004006       Technology   
9989  South  FUR-FU-10001889        Furniture   

                                           Nome_Produto  Valor_Venda  \
9971  Boston Model 1800 Electric Pencil Sharpener, Gray      140.750   
9980                         KI Adjustable-Height Table       85.980   
9987  Logitech G430 Surround Sound Gaming Headset wi...       79.990   
9988                    Panasonic KX - TS880B Telephone      206.100   
9989                             Ultra Door Pull Handle       25.248   

      Quantidade  
9971         5.0  
9980         1.0  
9987         1.0  
9988         5.0  
9989         3.0